src/Mvc/Mvc.ViewFeatures/src/Infrastructure/DefaultTempDataSerializer.cs

```diff
@@ -3,30 +3,142 @@
 
 using System;
 using System.Collections.Generic;
-using Microsoft.Extensions.DependencyInjection;
+using System.Globalization;
+using System.Text.Json;
 
 namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Infrastructure
 {
     internal class DefaultTempDataSerializer : TempDataSerializer
     {
-        public override IDictionary<string, object> Deserialize(byte[] unprotectedData)
+        public override IDictionary<string, object> Deserialize(byte[] value)
         {
-            throw new InvalidOperationException(Core.Resources.FormatReferenceToNewtonsoftJsonRequired(
-                Resources.DeserializingTempData,
-                "Microsoft.AspNetCore.Mvc.NewtonsoftJson",
-                nameof(IMvcBuilder),
-                "AddNewtonsoftJson",
-                "ConfigureServices(...)"));
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            if (value.Length == 0)
+            {
+                return new Dictionary<string, object>();
+            }
+
+            using var jsonDocument = JsonDocument.Parse(value);
+            var rootElement = jsonDocument.RootElement;
+            return DeserializeDictionary(rootElement);
+        }
+
+        private IDictionary<string, object> DeserializeDictionary(JsonElement rootElement)
+        {
+            var deserialized = new Dictionary<string, object>(StringComparer.Ordinal);
+
+            foreach (var item in rootElement.EnumerateObject())
```

Comment:
We're going to have to keep this code, so it's likely the deserializer that would change. This does have fewer features compared to the Json.Net one (it does not support arrays or dictionaries), however round-tripping is really broken with the current implementation (https://github.com/aspnet/AspNetCore/issues/8873). For anything but the most trivial cases, you get InvalidCastException. So doing a more limited set of types that we can correctly support all the time seems like a better solution for now.

Reply:
I think as long as the error message is reasonable for preview 4 this is the right thing to do.

I expected us to do something specialized here anyway since we want to serialize each entry specially.